### PR TITLE
plan: get req prop and phys plans together

### PR DESCRIPTION
**Pull request by hanfei1991**
After this PR, we attach `requiredProp` info to physical plans. So we can avoid call `Copy`.

### CODE REVIEW

**Comment by winoros** [comment]
In fact, these `Copy` are not needless. This pr just enlarge the generating space?

**Comment by hanfei1991** [comment]
@winoros yes

File: plan/plan.go [hunk 1]
```diff
@@ -195,20 +196,27 @@ type PhysicalPlan interface {
...
 	basePlan *basePlan
-	taskMap  map[string]task
+
+	taskMap map[string]task
 }
 
 type basePhysicalPlan struct {
 	basePlan *basePlan
+
+	reqProps []*requiredProp
```
Comment (winoros) [inline plan/plan.go]: add comment. or `childrenReqProps` is better?

File: plan/gen_physical_plans.go [hunk 2]
```diff
@@ -80,8 +82,32 @@ func (p *LogicalJoin) getEqAndOtherCondsByOffsets(offsets []int) ([]*expression.
...
 }
 
-func (p *LogicalJoin) getMergeJoin() []PhysicalPlan {
+// Only if the input required prop is the prefix fo join keys, we can pass through this property.
+func (p *PhysicalMergeJoin) tryToGetChildReqProp(prop *requiredProp) ([]*requiredProp, bool) {
+	p.expectedCnt = prop.expectedCnt
+	lProp := &requiredProp{taskTp: rootTaskType, cols: p.leftKeys, expectedCnt: math.MaxFloat64}
+	rProp := &requiredProp{taskTp: rootTaskType, cols: p.rightKeys, expectedCnt: math.MaxFloat64}
+	if !prop.isEmpty() {
+		if prop.desc {
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply plan/gen_physical_plans.go]: Yes, desc is always slow, we prune this case.

File: plan/plan.go (hunk 1 above)
Comment (hanfei1991) [reply plan/plan.go]: fixed

File: plan/gen_physical_plans.go [hunk 3]
```diff
@@ -80,8 +82,33 @@ func (p *LogicalJoin) getEqAndOtherCondsByOffsets(offsets []int) ([]*expression.
...
+			return nil, false
+		}
+	}
+
+	return []*requiredProp{lProp, rProp}, true
+}
+
+func (p *LogicalJoin) getMergeJoin(prop *requiredProp) []PhysicalPlan {
 	joins := make([]PhysicalPlan, 0, len(p.leftProperties))
+	// The leftProperties caches all the possible properties that be provided by its children.
```
Comment (winoros) [inline plan/gen_physical_plans.go]: be -> is.

File: plan/gen_physical_plans.go [hunk 4]
```diff
@@ -126,21 +156,24 @@ func (p *LogicalJoin) getHashSemiJoin() PhysicalPlan {
 	return semiJoin
 }
 
-func (p *LogicalJoin) getHashJoins() []PhysicalPlan {
+func (p *LogicalJoin) getHashJoins(prop *requiredProp) []PhysicalPlan {
+	if !prop.isEmpty() { // hash join doesn't promise any orders
```
Comment (winoros) [inline plan/gen_physical_plans.go]: add `.`, hash -> Hash

File: plan/initialize.go [hunk 5]
```diff
@@ -356,10 +369,9 @@ func (p PhysicalIndexReader) init(ctx context.Context) *PhysicalIndexReader {
 	return &p
 }
 
-func (p PhysicalIndexJoin) init(ctx context.Context, children ...Plan) *PhysicalIndexJoin {
+func (p PhysicalIndexJoin) init(ctx context.Context) *PhysicalIndexJoin {
```
Comment (shenli) [inline plan/initialize.go]: Miss props as the second parameter?

File: plan/gen_physical_plans.go [hunk 6]
```diff
@@ -308,67 +352,133 @@ func (p *LogicalJoin) tryToGetIndexJoin() ([]PhysicalPlan, bool) {
...
 	}
 	joins = append(joins, hashJoins...)
 	return joins
 }
 
-func (p *LogicalProjection) generatePhysicalPlans() []PhysicalPlan {
+// tryToGetChildProp will check if this sort property can be pushed or not.
+// When a sort column will be replaced by scalar function, we refuse it.
+// When a sort column will be replaced by a constant, we just remove it.
+func (p *LogicalProjection) tryToGetChildProp(prop *requiredProp) (*requiredProp, bool) {
```
Comment (shenli) [inline plan/gen_physical_plans.go]: The function name is not proper.

Comment (hanfei1991) [reply plan/gen_physical_plans.go]: The returning value is used for child's req prop.

File: plan/gen_physical_plans.go [hunk 7]
```diff
@@ -184,8 +220,12 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
 	return matchOffsets
 }
 
-func (p *LogicalJoin) constructIndexJoin(innerJoinKeys, outerJoinKeys []*expression.Column, outerIdx int, innerPlan PhysicalPlan) []PhysicalPlan {
+func (p *LogicalJoin) constructIndexJoin(prop *requiredProp, innerJoinKeys, outerJoinKeys []*expression.Column, outerIdx int, innerPlan PhysicalPlan) []PhysicalPlan {
 	joinType := p.JoinType
+	outerSchema := p.children[outerIdx].Schema()
+	if !prop.allColsFromSchema(outerSchema) {
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply plan/gen_physical_plans.go]: select * from s join t order by t.id, If join is index join and t is inner table. Join can't pass the order.

File: plan/physical_plan_builder.go [hunk 8]
```diff
@@ -83,15 +83,17 @@ func (p *baseLogicalPlan) convert2NewPhysicalPlan(prop *requiredProp) (t task, e
 	}
 	// Now we only consider rootTask.
 	if len(p.basePlan.children) == 0 {
+		if !prop.isEmpty() {
```
Comment (zimulala) [inline plan/physical_plan_builder.go]: Add comments here. Here puzzled me why we can't put the non-empty `prop` to `genPhysPlansByReqProp`.
And we can judge `prop.isEmpty()` first.

File: plan/gen_physical_plans.go [hunk 9]
```diff
@@ -308,84 +351,150 @@ func (p *LogicalJoin) tryToGetIndexJoin() ([]PhysicalPlan, bool) {
...
+		}.init(p.ctx, resultProp)
 		limit.profile = p.profile
-		plans = append(plans, limit)
+		limit.SetSchema(p.schema)
+		ret = append(ret, limit)
+	}
+	return ret
+}
+
+func itemsMatchProp(items []*ByItems, prop *requiredProp) bool {
```
Comment (zimulala) [inline plan/gen_physical_plans.go]: Could we put this function next to `isPrefix`? These two functions are similar.

File: plan/gen_physical_plans.go [hunk 10]
```diff
@@ -404,112 +513,152 @@ func (p *LogicalAggregation) getStreamAggs() []PhysicalPlan {
...
-		partial: p.partial,
-	}.init(p.ctx)
-	limit.profile = p.profile
-	limit.SetSchema(p.Schema())
-	return []PhysicalPlan{limit}
+func (p *LogicalLimit) genPhysPlansByReqProp(prop *requiredProp) []PhysicalPlan {
+	if !prop.isEmpty() {
+		return nil
+	}
+	ret := make([]PhysicalPlan, 0, 3)
```
Comment (zimulala) [inline plan/gen_physical_plans.go]: I think `len(wholeTaskTypes)` is better.

File: plan/gen_physical_plans.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zimulala) [reply plan/gen_physical_plans.go]: Please add the comment.
Rest LGTM

**Comment by hanfei1991** [comment]
@zimulala comment has been added

**Comment by zimulala** [comment]
LGTM

**Pull request merged**